# 유방암 환자 데이터셋을 활용한 지도학습 학습모델 평가

In [8]:
# warning 무시
import warnings
warnings.filterwarnings('ignore')

import matplotlib.pylab as plt
# '-' 깨짐 해결 모듈
import matplotlib as mpl

plt.rcParams["font.family"] = "Malgun Gothic"  # For Windows
plt.rcParams["font.size"] = 12
plt.rcParams["figure.figsize"] = (8,4)
print(plt.rcParams["font.family"])

# 마이너스 깨짐 해결
mpl.rcParams["axes.unicode_minus"] = False

['Malgun Gothic']


## 학습모델 평가 방법(Evaluation)
- Accuracy(정확도)
- Confusion Matrix(오차행렬)
- Precision(정밀도)
- Recall(재현율)
- F1 score
- ROC/AUC

In [24]:
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
import numpy as np
import pandas as pd
from sklearn.tree import DecisionTreeClassifier

유방암 환자 데이터셋 로드  
sklearn의 내장 dataset 활용 : load_breast_cancer  
target:
- 1 양성종양 : 착한 종양(Positive)
- 0 악성종양 : 나쁜 종양(Negative)


In [3]:
cancer = load_breast_cancer()

In [11]:
cancer
# data 값의 대응하는 target : y값

{'data': array([[1.799e+01, 1.038e+01, 1.228e+02, ..., 2.654e-01, 4.601e-01,
         1.189e-01],
        [2.057e+01, 1.777e+01, 1.329e+02, ..., 1.860e-01, 2.750e-01,
         8.902e-02],
        [1.969e+01, 2.125e+01, 1.300e+02, ..., 2.430e-01, 3.613e-01,
         8.758e-02],
        ...,
        [1.660e+01, 2.808e+01, 1.083e+02, ..., 1.418e-01, 2.218e-01,
         7.820e-02],
        [2.060e+01, 2.933e+01, 1.401e+02, ..., 2.650e-01, 4.087e-01,
         1.240e-01],
        [7.760e+00, 2.454e+01, 4.792e+01, ..., 0.000e+00, 2.871e-01,
         7.039e-02]]),
 'target': array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
        0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0,
        1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0,
        1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1,
        1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0

In [5]:
cancer.keys()

dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename', 'data_module'])

In [7]:
cancer.feature_names

array(['mean radius', 'mean texture', 'mean perimeter', 'mean area',
       'mean smoothness', 'mean compactness', 'mean concavity',
       'mean concave points', 'mean symmetry', 'mean fractal dimension',
       'radius error', 'texture error', 'perimeter error', 'area error',
       'smoothness error', 'compactness error', 'concavity error',
       'concave points error', 'symmetry error',
       'fractal dimension error', 'worst radius', 'worst texture',
       'worst perimeter', 'worst area', 'worst smoothness',
       'worst compactness', 'worst concavity', 'worst concave points',
       'worst symmetry', 'worst fractal dimension'], dtype='<U23')

In [9]:
cancer.target

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0,
       1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0,
       1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1,
       1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0,
       0, 1, 0, 0, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1,
       1, 0, 1, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0,
       0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0,
       1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1,
       1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0,

In [10]:
cancer.target_names

array(['malignant', 'benign'], dtype='<U9')

In [13]:
print(cancer["DESCR"])

.. _breast_cancer_dataset:

Breast cancer wisconsin (diagnostic) dataset
--------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 569

    :Number of Attributes: 30 numeric, predictive attributes and the class

    :Attribute Information:
        - radius (mean of distances from center to points on the perimeter)
        - texture (standard deviation of gray-scale values)
        - perimeter
        - area
        - smoothness (local variation in radius lengths)
        - compactness (perimeter^2 / area - 1.0)
        - concavity (severity of concave portions of the contour)
        - concave points (number of concave portions of the contour)
        - symmetry
        - fractal dimension ("coastline approximation" - 1)

        The mean, standard error, and "worst" or largest (mean of the three
        worst/largest values) of these features were computed for each image,
        resulting in 30 features.  For instance, field 0 is Mean Radi

In [14]:
# feature(X), target(y), feature_names 분리
data = cancer["data"]
target = cancer["target"]
feature_names = cancer["feature_names"]

In [15]:
print(data)
print("-"*60)
print(target)
print("-"*60)
print(feature_names)

[[1.799e+01 1.038e+01 1.228e+02 ... 2.654e-01 4.601e-01 1.189e-01]
 [2.057e+01 1.777e+01 1.329e+02 ... 1.860e-01 2.750e-01 8.902e-02]
 [1.969e+01 2.125e+01 1.300e+02 ... 2.430e-01 3.613e-01 8.758e-02]
 ...
 [1.660e+01 2.808e+01 1.083e+02 ... 1.418e-01 2.218e-01 7.820e-02]
 [2.060e+01 2.933e+01 1.401e+02 ... 2.650e-01 4.087e-01 1.240e-01]
 [7.760e+00 2.454e+01 4.792e+01 ... 0.000e+00 2.871e-01 7.039e-02]]
------------------------------------------------------------
[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 1 0 0 0 0 0 0 0 0 1 0 1 1 1 1 1 0 0 1 0 0 1 1 1 1 0 1 0 0 1 1 1 1 0 1 0 0
 1 0 1 0 0 1 1 1 0 0 1 0 0 0 1 1 1 0 1 1 0 0 1 1 1 0 0 1 1 1 1 0 1 1 0 1 1
 1 1 1 1 1 1 0 0 0 1 0 0 1 1 1 0 0 1 0 1 0 0 1 0 0 1 1 0 1 1 0 1 1 1 1 0 1
 1 1 1 1 1 1 1 1 0 1 1 1 1 0 0 1 0 1 1 0 0 1 1 0 0 1 1 1 1 0 1 1 0 0 0 1 0
 1 0 1 1 1 0 1 1 0 0 1 0 0 0 0 1 0 0 0 1 0 1 0 1 1 0 1 0 0 0 0 1 1 0 0 1 1
 1 0 1 1 1 1 1 0 0 1 1 0 1 1 0 0 1 0 1 1 1 1 0 1 1 1 1 1 0 1 0 0 0 0 0 0 0
 0 0 0

##### 데이터 프레임으로 생성

In [16]:
df = pd.DataFrame(data=data, columns=feature_names)
df

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0          17.99         10.38          122.80     1001.0          0.11840   
1          20.57         17.77          132.90     1326.0          0.08474   
2          19.69         21.25          130.00     1203.0          0.10960   
3          11.42         20.38           77.58      386.1          0.14250   
4          20.29         14.34          135.10     1297.0          0.10030   
..           ...           ...             ...        ...              ...   
564        21.56         22.39          142.00     1479.0          0.11100   
565        20.13         28.25          131.20     1261.0          0.09780   
566        16.60         28.08          108.30      858.1          0.08455   
567        20.60         29.33          140.10     1265.0          0.11780   
568         7.76         24.54           47.92      181.0          0.05263   

     mean compactness  mean concavity  mean concave points  mean symmetry  \
0             0.27760         0.30010              0.14710         0.2419   
1             0.07864         0.08690              0.07017         0.1812   
2             0.15990         0.19740              0.12790         0.2069   
3             0.28390         0.24140              0.10520         0.2597   
4             0.13280         0.19800              0.10430         0.1809   
..                ...             ...                  ...            ...   
564           0.11590         0.24390              0.13890         0.1726   
565           0.10340         0.14400              0.09791         0.1752   
566           0.10230         0.09251              0.05302         0.1590   
567           0.27700         0.35140              0.15200         0.2397   
568           0.04362         0.00000              0.00000         0.1587   

     mean fractal dimension  ...  worst radius  worst texture  \
0                   0.07871  ...        25.380          17.33   
1                   0.05667  ...        24.990          23.41   
2                   0.05999  ...        23.570          25.53   
3                   0.09744  ...        14.910          26.50   
4                   0.05883  ...        22.540          16.67   
..                      ...  ...           ...            ...   
564                 0.05623  ...        25.450          26.40   
565                 0.05533  ...        23.690          38.25   
566                 0.05648  ...        18.980          34.12   
567                 0.07016  ...        25.740          39.42   
568                 0.05884  ...         9.456          30.37   

     worst perimeter  worst area  worst smoothness  worst compactness  \
0             184.60      2019.0           0.16220            0.66560   
1             158.80      1956.0           0.12380            0.18660   
2             152.50      1709.0           0.14440            0.42450   
3              98.87       567.7           0.20980            0.86630   
4             152.20      1575.0           0.13740            0.20500   
..               ...         ...               ...                ...   
564           166.10      2027.0           0.14100            0.21130   
565           155.00      1731.0           0.11660            0.19220   
566           126.70      1124.0           0.11390            0.30940   
567           184.60      1821.0           0.16500            0.86810   
568            59.16       268.6           0.08996            0.06444   

     worst concavity  worst concave points  worst symmetry  \
0             0.7119                0.2654          0.4601   
1             0.2416                0.1860          0.2750   
2             0.4504                0.2430          0.3613   
3             0.6869                0.2575          0.6638   
4             0.4000                0.1625          0.2364   
..               ...                   ...             ...   
564           0.4107                0.2216         

In [17]:
df["target"] = cancer["target"]
df

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0          17.99         10.38          122.80     1001.0          0.11840   
1          20.57         17.77          132.90     1326.0          0.08474   
2          19.69         21.25          130.00     1203.0          0.10960   
3          11.42         20.38           77.58      386.1          0.14250   
4          20.29         14.34          135.10     1297.0          0.10030   
..           ...           ...             ...        ...              ...   
564        21.56         22.39          142.00     1479.0          0.11100   
565        20.13         28.25          131.20     1261.0          0.09780   
566        16.60         28.08          108.30      858.1          0.08455   
567        20.60         29.33          140.10     1265.0          0.11780   
568         7.76         24.54           47.92      181.0          0.05263   

     mean compactness  mean concavity  mean concave points  mean symmetry  \
0             0.27760         0.30010              0.14710         0.2419   
1             0.07864         0.08690              0.07017         0.1812   
2             0.15990         0.19740              0.12790         0.2069   
3             0.28390         0.24140              0.10520         0.2597   
4             0.13280         0.19800              0.10430         0.1809   
..                ...             ...                  ...            ...   
564           0.11590         0.24390              0.13890         0.1726   
565           0.10340         0.14400              0.09791         0.1752   
566           0.10230         0.09251              0.05302         0.1590   
567           0.27700         0.35140              0.15200         0.2397   
568           0.04362         0.00000              0.00000         0.1587   

     mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
0                   0.07871  ...          17.33           184.60      2019.0   
1                   0.05667  ...          23.41           158.80      1956.0   
2                   0.05999  ...          25.53           152.50      1709.0   
3                   0.09744  ...          26.50            98.87       567.7   
4                   0.05883  ...          16.67           152.20      1575.0   
..                      ...  ...            ...              ...         ...   
564                 0.05623  ...          26.40           166.10      2027.0   
565                 0.05533  ...          38.25           155.00      1731.0   
566                 0.05648  ...          34.12           126.70      1124.0   
567                 0.07016  ...          39.42           184.60      1821.0   
568                 0.05884  ...          30.37            59.16       268.6   

     worst smoothness  worst compactness  worst concavity  \
0             0.16220            0.66560           0.7119   
1             0.12380            0.18660           0.2416   
2             0.14440            0.42450           0.4504   
3             0.20980            0.86630           0.6869   
4             0.13740            0.20500           0.4000   
..                ...                ...              ...   
564           0.14100            0.21130           0.4107   
565           0.11660            0.19220           0.3215   
566           0.11390            0.30940           0.3403   
567           0.16500            0.86810           0.9387   
568           0.08996            0.06444           0.0000   

     worst concave points  worst symmetry  worst fractal dimension  target  
0                  0.2654          0.4601                  0.11890       0  
1                  0.1860          0.2750                  0.08902       0  
2                  0.2430          0.3613                  0.08758       0  
3                  0.2575          0.6638                  0.17300       0  
4                  0.1625          0.2364                  0.076

In [18]:
df_pos = df.loc[df["target"] == 1]
df_neg = df.loc[df["target"] == 0]

In [19]:
df_pos

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
19        13.540         14.36           87.46      566.3          0.09779   
20        13.080         15.71           85.63      520.0          0.10750   
21         9.504         12.44           60.34      273.9          0.10240   
37        13.030         18.42           82.61      523.8          0.08983   
46         8.196         16.84           51.71      201.9          0.08600   
..           ...           ...             ...        ...              ...   
558       14.590         22.68           96.39      657.1          0.08473   
559       11.510         23.93           74.52      403.5          0.09261   
560       14.050         27.15           91.38      600.4          0.09929   
561       11.200         29.37           70.67      386.0          0.07449   
568        7.760         24.54           47.92      181.0          0.05263   

     mean compactness  mean concavity  mean concave points  mean symmetry  \
19            0.08129         0.06664             0.047810         0.1885   
20            0.12700         0.04568             0.031100         0.1967   
21            0.06492         0.02956             0.020760         0.1815   
37            0.03766         0.02562             0.029230         0.1467   
46            0.05943         0.01588             0.005917         0.1769   
..                ...             ...                  ...            ...   
558           0.13300         0.10290             0.037360         0.1454   
559           0.10210         0.11120             0.041050         0.1388   
560           0.11260         0.04462             0.043040         0.1537   
561           0.03558         0.00000             0.000000         0.1060   
568           0.04362         0.00000             0.000000         0.1587   

     mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
19                  0.05766  ...          19.26            99.70       711.2   
20                  0.06811  ...          20.49            96.09       630.5   
21                  0.06905  ...          15.66            65.13       314.9   
37                  0.05863  ...          22.81            84.46       545.9   
46                  0.06503  ...          21.96            57.26       242.2   
..                      ...  ...            ...              ...         ...   
558                 0.06147  ...          27.27           105.90       733.5   
559                 0.06570  ...          37.16            82.28       474.2   
560                 0.06171  ...          33.17           100.20       706.7   
561                 0.05502  ...          38.30            75.19       439.6   
568                 0.05884  ...          30.37            59.16       268.6   

     worst smoothness  worst compactness  worst concavity  \
19            0.14400            0.17730          0.23900   
20            0.13120            0.27760          0.18900   
21            0.13240            0.11480          0.08867   
37            0.09701            0.04619          0.04833   
46            0.12970            0.13570          0.06880   
..                ...                ...              ...   
558           0.10260            0.31710          0.36620   
559           0.12980            0.25170          0.36300   
560           0.12410            0.22640          0.13260   
561           0.09267            0.05494          0.00000   
568           0.08996            0.06444          0.00000   

     worst concave points  worst symmetry  worst fractal dimension  target  
19                0.12880          0.2977                  0.07259       1  
20                0.07283          0.3184                  0.08183       1  
21                0.06227          0.2450                  0.07773       1  
37                0.05013          0.1987                  0.06169       1  
46                0.02564          0.3105                  0.074

In [20]:
df_neg

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0          17.99         10.38          122.80     1001.0          0.11840   
1          20.57         17.77          132.90     1326.0          0.08474   
2          19.69         21.25          130.00     1203.0          0.10960   
3          11.42         20.38           77.58      386.1          0.14250   
4          20.29         14.34          135.10     1297.0          0.10030   
..           ...           ...             ...        ...              ...   
563        20.92         25.09          143.00     1347.0          0.10990   
564        21.56         22.39          142.00     1479.0          0.11100   
565        20.13         28.25          131.20     1261.0          0.09780   
566        16.60         28.08          108.30      858.1          0.08455   
567        20.60         29.33          140.10     1265.0          0.11780   

     mean compactness  mean concavity  mean concave points  mean symmetry  \
0             0.27760         0.30010              0.14710         0.2419   
1             0.07864         0.08690              0.07017         0.1812   
2             0.15990         0.19740              0.12790         0.2069   
3             0.28390         0.24140              0.10520         0.2597   
4             0.13280         0.19800              0.10430         0.1809   
..                ...             ...                  ...            ...   
563           0.22360         0.31740              0.14740         0.2149   
564           0.11590         0.24390              0.13890         0.1726   
565           0.10340         0.14400              0.09791         0.1752   
566           0.10230         0.09251              0.05302         0.1590   
567           0.27700         0.35140              0.15200         0.2397   

     mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
0                   0.07871  ...          17.33           184.60      2019.0   
1                   0.05667  ...          23.41           158.80      1956.0   
2                   0.05999  ...          25.53           152.50      1709.0   
3                   0.09744  ...          26.50            98.87       567.7   
4                   0.05883  ...          16.67           152.20      1575.0   
..                      ...  ...            ...              ...         ...   
563                 0.06879  ...          29.41           179.10      1819.0   
564                 0.05623  ...          26.40           166.10      2027.0   
565                 0.05533  ...          38.25           155.00      1731.0   
566                 0.05648  ...          34.12           126.70      1124.0   
567                 0.07016  ...          39.42           184.60      1821.0   

     worst smoothness  worst compactness  worst concavity  \
0              0.1622             0.6656           0.7119   
1              0.1238             0.1866           0.2416   
2              0.1444             0.4245           0.4504   
3              0.2098             0.8663           0.6869   
4              0.1374             0.2050           0.4000   
..                ...                ...              ...   
563            0.1407             0.4186           0.6599   
564            0.1410             0.2113           0.4107   
565            0.1166             0.1922           0.3215   
566            0.1139             0.3094           0.3403   
567            0.1650             0.8681           0.9387   

     worst concave points  worst symmetry  worst fractal dimension  target  
0                  0.2654          0.4601                  0.11890       0  
1                  0.1860          0.2750                  0.08902       0  
2                  0.2430          0.3613                  0.08758       0  
3                  0.2575          0.6638                  0.17300       0  
4                  0.1625          0.2364                  0.076

In [21]:
print(df_pos.shape, df_neg.shape)

(357, 31) (212, 31)


## input 데이터 준비

In [22]:
X = df.drop("target", axis=1) # features
y = df["target"]

## train, test 데이터 분할

In [23]:
# test_size = 0.25,  shuffle=True(default)
train_X, test_X, train_y, test_y = train_test_split(X, y, random_state=20, stratify=y, test_size=0.25, shuffle=True)

## 모델 생성 및 학습

In [25]:
model_dtc = DecisionTreeClassifier(max_depth=5)
model_dtc.fit(train_X, train_y)

DecisionTreeClassifier(max_depth=5)

## 학습 모델 테스트

In [27]:
y_pred = model_dtc.predict(test_X)
y_pred

array([1, 1, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1,
       1, 1, 0, 1, 0, 1, 1, 1, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0,
       1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 1, 1, 0, 1, 0, 0,
       1, 1, 1, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1,
       1, 1, 0, 1, 1, 0, 0, 1, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 0, 0, 0, 1,
       1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1])

## 성능 평가

In [29]:
print("모델의 정확도 : %.2f" % (y_pred == test_y).mean())

모델의 정확도 : 0.93


### 정확도 확인 함수 sklearn.metrics.accuracy_score(y_true, y_pred)
- sklearn.metrics.accuracy_score(y_true, y_pred)
- https://scikit-learn.org/stable/modules/generated/sklearn.metrics.accuracy_score.html?highlight=accuracy_score#sklearn.metrics.accuracy_score

In [30]:
from sklearn.metrics import accuracy_score
print("모델의 정확도 : %.2f" % accuracy_score(y_pred, test_y))

모델의 정확도 : 0.93


In [33]:
# 테스트 데이터(X_test)와 labels(y_test)로 모델의 정확도를 계산
print("모델의 정확도 : %.2f" % model_dtc.score(test_X, test_y))

모델의 정확도 : 0.93


## 성능평가 함수 설명
- from sklearn.metrics import classification_report
- https://scikit-learn.org/stable/modules/generated/sklearn.metrics.classification_report.html?highlight=classification_report

In [34]:
from sklearn.metrics import classification_report

In [35]:
# Predict를 수행하고 classification_report() 결과 출력
y_pred = model_dtc.predict(test_X)
print(classification_report(y_pred, test_y, target_names=cancer.target_names))

              precision    recall  f1-score   support

   malignant       0.94      0.88      0.91        57
      benign       0.92      0.97      0.94        86

    accuracy                           0.93       143
   macro avg       0.93      0.92      0.93       143
weighted avg       0.93      0.93      0.93       143



In [37]:
print(classification_report(y_pred, test_y, labels=[0, 1]))

              precision    recall  f1-score   support

           0       0.94      0.88      0.91        57
           1       0.92      0.97      0.94        86

    accuracy                           0.93       143
   macro avg       0.93      0.92      0.93       143
weighted avg       0.93      0.93      0.93       143



In [38]:
print(classification_report(y_pred, test_y))

              precision    recall  f1-score   support

           0       0.94      0.88      0.91        57
           1       0.92      0.97      0.94        86

    accuracy                           0.93       143
   macro avg       0.93      0.92      0.93       143
weighted avg       0.93      0.93      0.93       143



#### macro avg, weighted avg라는 값이 출력이 된다. 각각 아래와같은 의미를 가진다

- macro avg:
    - 각 클래스에 대한 f1-score(precision, recall)를 계산한 뒤 단순한 산술 평균을 내어 값을 계산함.
- weighted avg:
    - 불균형한 데이터일 경우에 보는 수치값.
    - 각 class에 해당하는 data의 개수에 가중치를 주어 평균을 구한 것

출처

- https://jaeyung1001.tistory.com/252
- https://scikit-learn.org/stable/modules/generated/sklearn.metrics.classification_report.html?highlight=classification_report#sklearn.metrics.classification_report

## 양성 환자 편향 데이터로 구성

극단적으로 데이터 표현, 다음과 같이 data 생성  
양성 환자 **357개** + 악성 환자 **5개**

In [39]:
df_pos.head(5)

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
19       13.540         14.36           87.46      566.3          0.09779   
20       13.080         15.71           85.63      520.0          0.10750   
21        9.504         12.44           60.34      273.9          0.10240   
37       13.030         18.42           82.61      523.8          0.08983   
46        8.196         16.84           51.71      201.9          0.08600   

    mean compactness  mean concavity  mean concave points  mean symmetry  \
19           0.08129         0.06664             0.047810         0.1885   
20           0.12700         0.04568             0.031100         0.1967   
21           0.06492         0.02956             0.020760         0.1815   
37           0.03766         0.02562             0.029230         0.1467   
46           0.05943         0.01588             0.005917         0.1769   

    mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
19                 0.05766  ...          19.26            99.70       711.2   
20                 0.06811  ...          20.49            96.09       630.5   
21                 0.06905  ...          15.66            65.13       314.9   
37                 0.05863  ...          22.81            84.46       545.9   
46                 0.06503  ...          21.96            57.26       242.2   

    worst smoothness  worst compactness  worst concavity  \
19           0.14400            0.17730          0.23900   
20           0.13120            0.27760          0.18900   
21           0.13240            0.11480          0.08867   
37           0.09701            0.04619          0.04833   
46           0.12970            0.13570          0.06880   

    worst concave points  worst symmetry  worst fractal dimension  target  
19               0.12880          0.2977                  0.07259       1  
20               0.07283          0.3184                  0.08183       1  
21               0.06227          0.2450                  0.07773       1  
37               0.05013          0.1987                  0.06169       1  
46               0.02564          0.3105                  0.07409       1  

[5 rows x 31 columns]

In [40]:
df_neg[:5]

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
0                 0.07871  ...          17.33           184.60      2019.0   
1                 0.05667  ...          23.41           158.80      1956.0   
2                 0.05999  ...          25.53           152.50      1709.0   
3                 0.09744  ...          26.50            98.87       567.7   
4                 0.05883  ...          16.67           152.20      1575.0   

   worst smoothness  worst compactness  worst concavity  worst concave points  \
0            0.1622             0.6656           0.7119                0.2654   
1            0.1238             0.1866           0.2416                0.1860   
2            0.1444             0.4245           0.4504                0.2430   
3            0.2098             0.8663           0.6869                0.2575   
4            0.1374             0.2050           0.4000                0.1625   

   worst symmetry  worst fractal dimension  target  
0          0.4601                  0.11890       0  
1          0.2750                  0.08902       0  
2          0.3613                  0.08758       0  
3          0.6638                  0.17300       0  
4          0.2364                  0.07678       0  

[5 rows x 31 columns]

In [42]:
df_pos

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
19        13.540         14.36           87.46      566.3          0.09779   
20        13.080         15.71           85.63      520.0          0.10750   
21         9.504         12.44           60.34      273.9          0.10240   
37        13.030         18.42           82.61      523.8          0.08983   
46         8.196         16.84           51.71      201.9          0.08600   
..           ...           ...             ...        ...              ...   
558       14.590         22.68           96.39      657.1          0.08473   
559       11.510         23.93           74.52      403.5          0.09261   
560       14.050         27.15           91.38      600.4          0.09929   
561       11.200         29.37           70.67      386.0          0.07449   
568        7.760         24.54           47.92      181.0          0.05263   

     mean compactness  mean concavity  mean concave points  mean symmetry  \
19            0.08129         0.06664             0.047810         0.1885   
20            0.12700         0.04568             0.031100         0.1967   
21            0.06492         0.02956             0.020760         0.1815   
37            0.03766         0.02562             0.029230         0.1467   
46            0.05943         0.01588             0.005917         0.1769   
..                ...             ...                  ...            ...   
558           0.13300         0.10290             0.037360         0.1454   
559           0.10210         0.11120             0.041050         0.1388   
560           0.11260         0.04462             0.043040         0.1537   
561           0.03558         0.00000             0.000000         0.1060   
568           0.04362         0.00000             0.000000         0.1587   

     mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
19                  0.05766  ...          19.26            99.70       711.2   
20                  0.06811  ...          20.49            96.09       630.5   
21                  0.06905  ...          15.66            65.13       314.9   
37                  0.05863  ...          22.81            84.46       545.9   
46                  0.06503  ...          21.96            57.26       242.2   
..                      ...  ...            ...              ...         ...   
558                 0.06147  ...          27.27           105.90       733.5   
559                 0.06570  ...          37.16            82.28       474.2   
560                 0.06171  ...          33.17           100.20       706.7   
561                 0.05502  ...          38.30            75.19       439.6   
568                 0.05884  ...          30.37            59.16       268.6   

     worst smoothness  worst compactness  worst concavity  \
19            0.14400            0.17730          0.23900   
20            0.13120            0.27760          0.18900   
21            0.13240            0.11480          0.08867   
37            0.09701            0.04619          0.04833   
46            0.12970            0.13570          0.06880   
..                ...                ...              ...   
558           0.10260            0.31710          0.36620   
559           0.12980            0.25170          0.36300   
560           0.12410            0.22640          0.13260   
561           0.09267            0.05494          0.00000   
568           0.08996            0.06444          0.00000   

     worst concave points  worst symmetry  worst fractal dimension  target  
19                0.12880          0.2977                  0.07259       1  
20                0.07283          0.3184                  0.08183       1  
21                0.06227          0.2450                  0.07773       1  
37                0.05013          0.1987                  0.06169       1  
46                0.02564          0.3105                  0.074

In [41]:
df_sample = pd.concat([df_pos, df_neg[:5]], sort=True)
df_sample

area error  compactness error  concave points error  concavity error  \
19      23.560           0.014600              0.013150          0.02387   
20      14.670           0.018980              0.006490          0.01698   
21      15.700           0.014320              0.014210          0.01985   
37      14.160           0.004899              0.011640          0.01343   
46       8.205           0.016460              0.005917          0.01588   
..         ...                ...                   ...              ...   
0      153.400           0.049040              0.015870          0.05373   
1       74.080           0.013080              0.013400          0.01860   
2       94.030           0.040060              0.020580          0.03832   
3       27.230           0.074580              0.018670          0.05661   
4       94.440           0.024610              0.018850          0.05688   

    fractal dimension error  mean area  mean compactness  mean concave points  \
19                 0.002300      566.3           0.08129             0.047810   
20                 0.002425      520.0           0.12700             0.031100   
21                 0.002968      273.9           0.06492             0.020760   
37                 0.001777      523.8           0.03766             0.029230   
46                 0.002582      201.9           0.05943             0.005917   
..                      ...        ...               ...                  ...   
0                  0.006193     1001.0           0.27760             0.147100   
1                  0.003532     1326.0           0.07864             0.070170   
2                  0.004571     1203.0           0.15990             0.127900   
3                  0.009208      386.1           0.28390             0.105200   
4                  0.005115     1297.0           0.13280             0.104300   

    mean concavity  mean fractal dimension  ...  worst area  \
19         0.06664                 0.05766  ...       711.2   
20         0.04568                 0.06811  ...       630.5   
21         0.02956                 0.06905  ...       314.9   
37         0.02562                 0.05863  ...       545.9   
46         0.01588                 0.06503  ...       242.2   
..             ...                     ...  ...         ...   
0          0.30010                 0.07871  ...      2019.0   
1          0.08690                 0.05667  ...      1956.0   
2          0.19740                 0.05999  ...      1709.0   
3          0.24140                 0.09744  ...       567.7   
4          0.19800                 0.05883  ...      1575.0   

    worst compactness  worst concave points  worst concavity  \
19            0.17730               0.12880          0.23900   
20            0.27760               0.07283          0.18900   
21            0.11480               0.06227          0.08867   
37            0.04619               0.05013          0.04833   
46            0.13570               0.02564          0.06880   
..                ...                   ...              ...   
0             0.66560               0.26540          0.71190   
1             0.18660               0.18600          0.24160   
2             0.42450               0.24300          0.45040   
3             0.86630               0.25750          0.68690   
4             0.20500               0.16250          0.40000   

    worst fractal dimension  worst perimeter  worst radius  worst smoothness  \
19                  0.07259            99.70        15.110           0.14400   
20                  0.08183            96.09        14.500           0.13120   
21                  0.07773            65.13        10.230           0.13240   
37                  0.06169            84.46        13.300           0.09701   
46                  0.07409            57.26         8.964           0.12970   
..                      ...              ...           ...               ...   
0                   0.11890     

In [43]:
X = df_sample.drop("target", axis=1)
y = df_sample["target"]

In [45]:
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=2, shuffle=True, stratify=y)

## 모델 생성 및 모델 학습

In [46]:
model_dtc_sp = DecisionTreeClassifier(max_depth=5)
model_dtc_sp.fit(X_train, y_train)

DecisionTreeClassifier(max_depth=5)

In [48]:
pred_y = model_dtc_sp.predict(X_test)
pred_y

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1])

In [49]:
y_test.to_numpy()

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1])

In [51]:
(pred_y == y_test).mean()

1.0

In [53]:
print(classification_report(y_test, pred_y, target_names=cancer.target_names, zero_division=True))

              precision    recall  f1-score   support

   malignant       1.00      1.00      1.00         1
      benign       1.00      1.00      1.00        90

    accuracy                           1.00        91
   macro avg       1.00      1.00      1.00        91
weighted avg       1.00      1.00      1.00        91



#### 의사가 모든 암환자 진단을 양성으로 판단한다고 가정하면,

In [54]:
# 모두 양성종양으로 진단하는 데이터 구성
pred_temp = np.ones(shape=y_test.shape)
pred_temp

array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1.])

In [55]:
pred_temp == y_test

425    True
442    True
345    True
528    True
279    True
       ... 
342    True
429    True
288    True
136    True
178    True
Name: target, Length: 91, dtype: bool

In [56]:
(pred_temp == y_test).mean()

0.989010989010989

의사가 모두 양성종양으로 진단한것과 ML 예측을 비교했을 때,  
예측률은 높으나 유능한 의사라 보기 어려움

## 정확도의 모순
- 정확도(Accuracy) : 실제 데이터에서 예측 데이터가 얼마나 같은지 판단하는 자료
- 정확도(Accuracy) = (예측 결과와 실제 데이터가 동일한 건수) / (전체 예측 데이터 건수)

정확도(accuracy)만 보고 분류기의 성능을 판별하는 것은 **위와 같은 오류**에 빠질 수 있음.

이를 보완하기 위해 나온 지표들을 살펴보자.


### confusion maxtix(오차 행렬, 혼동 행렬)

이진 분류의 **예측 오류가 얼마인지와 어떠한 유형의 예측 오류**가 발생하고 있는지를 함께 나타내는 자료


<img src="http://www.info.univ-angers.fr/~gh/Predipath/confusion.png" width="1000" height="1000">

숫자 5인지, 아닌지를 맞추는 모델

<img src="https://img1.daumcdn.net/thumb/R1280x0/?scode=mtistory2&fname=https%3A%2F%2Fblog.kakaocdn.net%2Fdn%2FPPAoe%2FbtrewxCS6Jn%2FHNzbSntqGyhIcPvv1eNWX1%2Fimg.png" width="1000" height="1000">

<img src="https://img1.daumcdn.net/thumb/R1280x0/?scode=mtistory2&fname=https%3A%2F%2Fblog.kakaocdn.net%2Fdn%2FbKYE4M%2FbtrewgA9ioX%2F7geXmrx7gC8mAQKEFyQcgK%2Fimg.png" width="1000" height="1000">

정확도(Accuracy) : 실제 데이터에서 예측 데이터가 얼마나 같은지를 판단하는 자료

### 정확도(Accuracy) = (TN+TP) / (TN+FP+FN+TP)

In [57]:
# confusion_matrix 결과 확인
from sklearn.metrics import confusion_matrix